internal/provider/api_key_resource.go

```diff
@@ -0,0 +1,316 @@
+			},
+			"secret": schema.StringAttribute{
+				Computed:    true,
+				Required:    false,
+				Optional:    false,
+				Sensitive:   true,
+				Description: "The secret portion of the API key. This is only available when creating a new key.",
+			},
+		},
+		Blocks: map[string]schema.Block{
+			"permissions": schema.ListNestedBlock{
+				Description: "Permissions control what actions the API key can perform.",
+				Validators: []validator.List{
+					listvalidator.SizeAtMost(1),
+				},
+				PlanModifiers: []planmodifier.List{
+					listplanmodifier.RequiresReplace(),
+				},
+				NestedObject: schema.NestedBlockObject{
+					Attributes: map[string]schema.Attribute{
+						"create_datasets": schema.BoolAttribute{
+							Optional:    true,
+							Computed:    true,
+							Default:     booldefault.StaticBool(false),
+							Description: "Allow this key to create missing datasets when sending telemetry.",
+							PlanModifiers: []planmodifier.Bool{
+								boolplanmodifier.RequiresReplace(),
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (r *apiKeyResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var plan models.APIKeyResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	newKey := &v2client.APIKey{
+		Name:        plan.Name.ValueStringPointer(),
+		KeyType:     plan.Type.ValueString(),
+		Environment: &v2client.Environment{ID: plan.EnvironmentID.ValueString()},
+		Disabled:    plan.Disabled.ValueBoolPointer(),
+		Permissions: expandAPIKeyPermissions(ctx, plan.Permissions, &resp.Diagnostics),
+	}
+
+	key, err := r.client.APIKeys.Create(ctx, newKey)
+	if helper.AddDiagnosticOnError(&resp.Diagnostics, "Creating Honeycomb API Key", err) {
+		return
+	}
+
+	var state models.APIKeyResourceModel
+	state.ID = types.StringValue(key.ID)
+	state.Name = types.StringValue(*key.Name)
+	state.Type = types.StringValue(key.KeyType)
+	state.EnvironmentID = types.StringValue(key.Environment.ID)
+	state.Disabled = types.BoolValue(*key.Disabled)
+	state.Secret = types.StringValue(key.Secret)
+
+	if !plan.Permissions.IsNull() {
+		state.Permissions = flattenAPIKeyPermissions(ctx, key.Permissions, &resp.Diagnostics)
+	} else {
+		state.Permissions = types.ListNull(types.ObjectType{AttrTypes: models.APIKeyPermissionsAttrType})
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, state)...)
+}
+
+func (r *apiKeyResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var state models.APIKeyResourceModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var detailedErr hny.DetailedError
+	key, err := r.client.APIKeys.Get(ctx, state.ID.ValueString())
+	if errors.As(err, &detailedErr) {
+		if detailedErr.IsNotFound() {
+			// if not found consider it deleted -- so just remove it from state
+			resp.State.RemoveResource(ctx)
+			return
```

Comment:
Good question! The provider takes care of that on it's own because of the ways it calls these methods (Read is called during plan when a delete or update is the action)